### DISCUSSION TITLE
filer store distributed-distributed replication

### DISCUSSION THREAD

**Discussion by gfxlabs** [seaweedfs/seaweedfs#2724] [Q&A]
hi! i had a question

reading this article: https://github.com/chrislusf/seaweedfs/wiki/Filer-Store-Replication

states that topology
filer1(mysql) <-> filer2(mysql)  
is okay

however, 
filer1(leveldb) <--XX NOT WORKING XX---> filer2(mysql)
is not okay

right now my topology is

filer1(redis2) <-> filer2(redis2)  <->filer3(redis2)

i was wondering if having multiple different distributed stores would work
filer1(redis2) <-> filer2(redis2)  <-> filer3(postgresql)
or
filer1(redis2) <-> filer2(redis3)  <-> filer3(postgresql)

I assume that it would not given the examples. but i am not 100% sure, since I connected a redis3 filer to my redis2 filers and it seemed to sync, but im not sure if that is an edge case.

**Comment by gfxlabs** [comment]
this comes from the fact that I would like both to be able to query by prefix using postgresql, and also access in constant time via redis

if there's another way to do that it would be useful

my use case is i store a quadtree 13 layers deep, each directory holding up to 4 images and 4 directories. it powers a flat map!

I'd love to be able to list out the directory but it takes rather long time on redis2 since there's no text index.

example:
![image](https://user-images.githubusercontent.com/86091021/156389963-ea14bf16-27f0-4217-a4f8-271bbe31b6cb.png)
 
(images are 1-30kb)

**Reply by chrislusf** [reply]
what is the query pattern? listing all the directories in one shot, or list directories one by one?

**Reply by gfxlabs** [reply]
what currently happens is we query for every possible file within the region boundaries, (file names are deterministic) and simply ignore the 404 errors. here's an example (querying for 2/3/3/3/?/?/?/? and 2/3/3/4/?/?/?/? here)

![image](https://user-images.githubusercontent.com/86091021/156507279-e5aa1e28-2a68-4f12-9f83-f048f2071220.png)

in this case, we would want every file which exists that is in form 2/3/3/3/?/?/?/?.png and 2/3/3/4/?/?/?/?.png

the issue is right now as you see, sparse areas of the map basically just spam 404s. maybe the best solution here is to just create some other program that reads filer logs and keeps those indices?

**Reply by ddorian** [reply]
Just use Postgresql and don't complicate it. Should be fast enough. Is it not?

**Reply by gfxlabs** [reply]
> Just use Postgresql and don't complicate it. Should be fast enough. Is it not?

right, but if it was possible to run both, i see no reason not to. i have the storage + spare cycles required and being able to put many different things in one place is a bit easier development wise.

**Reply by ddorian** [reply]
You are overcomplicating things and wasting everybodys time trying to prematurely optimize something that you'll never need

**Reply by chrislusf** [reply]
Current SQL filer store can not do prefix queries, with the primary key format as `<parent_directory_hash, name>`.

The cassandra filers store data layout is good for these prefix queries, with key format as the full path to an file entry. `/full/path/to/entry`.

**Reply by ddorian** [reply]
> The cassandra filers store data layout is good for these prefix queries, with key format as the full path to an file entry. /full/path/to/entry.

I expected it to be the same in sql.

**Reply by gfxlabs** [reply]
gotcha, i'll look into cassandra and see how it is

in general then, would the system behave if i had two different distributed stores? or would i just need to migrate

**Reply by ddorian** [reply]
@chrislusf any specific reason it uses `<parent_directory_hash, name>` in SQL?

**Reply by chrislusf** [reply]
MySQL has a limit of 1000 when using string as a primary key, if I remember correctly.

**Reply by renweijun** [reply]
@gfxlabs 
filer1(redis2) <-> filer2(redis2) <-> filer3(postgresql)
or
filer1(redis2) <-> filer2(redis3) <-> filer3(postgresql)
is okay ?

**Reply by renweijun** [reply]
@chrislusf 
Two filers with  two different shared store eg:redis3 tikv or postgresql, The  filer.store.id is different ?

**Reply by chrislusf** [reply]
@renweijun what is the setup of other components?

**Reply by renweijun** [reply]
@chrislusf 
example:
3 master 
/seaweedfs/weed -v=4 master -ip=master1 -mdir=/data/seaweedfs_master_data -port=9333 -metrics.address=1.1.1.1:9292 -defaultReplication=010 -peers=master1,master2,master3

3 filer filer1(redis3) <-> filer2(redis3) <-> filer3(tikv)
/seaweedfs/weed -v=4 filer -s3 -s3.config=/data/app/seaweedfs/cfg/s3config.json -s3.port=8333 -port=8888 -master=master1,master2,master3

many volume server
/seaweedfs/weed -v=4 volume -max=0 -disk= -dir=/data/seaweedfs_data -mserver=master1,master2,master3 -dataCenter=xx -compactionMBps=2048 -port=9221 -port.public=9221 -rack=F-05 -minFreeSpacePercent=10
  
Other setup defaults.  im not sure  filer1(redis3) <-> filer2(redis3) <working? or not woring?> filer3(tikv) 

server1 mount  fuse - filer1,filer2 
server2 mount  fuse  filer3 ,if  write some files in server1, server2 can read?

**Reply by chrislusf** [reply]
if in the same cluster, the remote filer store is expected to be the same instance.

**Reply by renweijun** [reply]
Got it, Thanks!
